### PR TITLE
Issue11

### PR DESCRIPTION
**Pull request by alicerobson**
Hi Colin
Addresses issue #11 
Turning Ids into a set had the advantage that the sorting is more automatic and there is much less need for the WITHSORT flag.
I have made the Edge matrix so that it does not need to contain all combinations. We may want to consider this in Python at some point.
Note that I will need to make one or two v minor changes to  python so as to get a match. 
Note also that I was not able to come up with a templated getCollection. I can explain when we next speak. But I have made several improvements along the lines you suggested.

### CODE REVIEW

File: papas/datatypes/DefinitionsCollections.h [hunk 1]
```diff
@@ -15,6 +15,11 @@
 #include "papas/graphtools/Edge.h"
 #include <list>
 #include <unordered_map>
+#if WITHSORT
```
Comment (cbernet) [inline papas/datatypes/DefinitionsCollections.h]: cool

File: papas/reconstruction/src/BuildPFBlocks.cpp [hunk 2]
```diff
@@ -12,30 +12,39 @@
 
 namespace papas {
 
-void buildPFBlocks(const Event& event, const std::string& ecalTypeAndSubtype, const std::string& hcalTypeAndSubtype,
-                   char trackSubtype, Blocks& blocks, Nodes& history) {
+void buildPFBlocks(const Event& event, IdCoder::SubType ecalSubtype, IdCoder::SubType hcalSubtype, char trackSubtype,
```
Comment (cbernet) [inline papas/reconstruction/src/BuildPFBlocks.cpp]: That's indeed better.

File: papas/reconstruction/src/PFBlock.cpp [hunk 3]
```diff
@@ -27,17 +27,17 @@ bool blockIdComparer(Identifier id1, Identifier id2) {
...
   for (auto id1 : m_elementIds) {
     for (auto id2 : m_elementIds) {
       if (id1 >= id2) continue;
       // move the edge from one unordered map to the other
-      auto e = edges.find(Edge::makeKey(id1, id2));
-      if (e == edges.end()) throw std::range_error("Required Edge is missing from edges collection");
-      m_edges.emplace(e->second.key(), std::move(e->second));
-      edges.erase(e);
+      const auto& e = edges.find(Edge::makeKey(id1, id2));
+      if (e != edges.end()) {
```
Comment (cbernet) [inline papas/reconstruction/src/PFBlock.cpp]: you're not throwing an exception anymore. Are you sure that's find? What if the edge is not found?

File: papas/reconstruction/src/PFBlock.cpp [hunk 4]
```diff
@@ -68,12 +68,20 @@ std::string PFBlock::shortName() const {
...
 
-const Edge& PFBlock::findEdge(Edge::EdgeKey key) const {
+const Edge& PFBlock::edge(Edge::EdgeKey key) const {
   auto edge = m_edges.find(key);
   if (edge == m_edges.end()) throw std::range_error("Edge not found");
   return edge->second;
 }
 
+const Edge& PFBlock::edge(Identifier id1, Identifier id2) const {
+  /// Find the edge corresponding to e1 e2
```
Comment (cbernet) [inline papas/reconstruction/src/PFBlock.cpp]: put that as a doxygen docstring in the .h and remove. 
no need to get into details, just say that the order of the ids does not matter.

File: papas/reconstruction/src/PFBlock.cpp [hunk 5]
```diff
@@ -162,28 +169,34 @@ std::string PFBlock::edgeMatrixString() const {
...
           out.write("       .");  // diagonal
           break;
-        } else if (edge(e1, e2).distance() < 0)
-          out.write("     ---");  //-ve distance
-        else if (edge(e1, e2).isLinked() == false)
-          out.write("     xxx");  // not linked
-        else {                    // linked and has distance
-          out.write("{:8.4f}", edge(e1, e2).distance());
+        } else if (hasEdge(e1, e2)) {
+          const auto& ed = edge(e1, e2);
```
Comment (cbernet) [inline papas/reconstruction/src/PFBlock.cpp]: just a detail but: in that case I would go for the explicit "edge" as a variable name rather that "ed"

File: papas/reconstruction/src/PFBlock.cpp (hunk 3 above)
Comment (alicerobson) [inline papas/reconstruction/src/PFBlock.cpp]: Hi Colin 
This is the unordered_map find.
I check in the following line to see if it was found ( it equals edges.end() if not found).
If the edge is not found its not added to the PFBlock edges collection. 
Papas should work fine provided we have an edge for everything that is linked. In fact, as we only ever remove links we could probably simplify this even further, and only have an edge if there is a link, in which case an edge need only contain the distance... A little more thought needed though.

File: papas/reconstruction/src/PFBlock.cpp (hunk 4 above)
Comment (alicerobson) [inline papas/reconstruction/src/PFBlock.cpp]: Done

File: papas/reconstruction/src/PFBlock.cpp (hunk 5 above)
Comment (alicerobson) [inline papas/reconstruction/src/PFBlock.cpp]: Done. Note I had to quality the edge function to PFBlock::edge to get it to accept  edge as the variable name.. const auto& edge = PFBlock::edge(e1, e2);

**Pull request merged**